### PR TITLE
Bug 1224855 - Pasteboard image fixes

### PR DESCRIPTION
**Pull request by thebnich**
Looking at our Copy Image code, I see a few different issues:
 1. In the Alamofire response, we do `application.endBackgroundTask()` only if the pasteboard string equals the copied string. We do this check to prevent replacing the pasteboard contents in case it's changed after any long-running image fetching, but ending the background task should be done unconditionally.
 2. `UIImage(data: NSData)` can return `nil` if the image creation fails for whatever reason, and we don't guard against this. I tested explicitly assigning `pasteBoard.image = nil`, and that caused the same crash reported in the bug, so I'm guessing that's the issue. Not sure why iOS allows setting a `nil` image if it's not supported!
 3. Setting the image replaces the image URL text in the pasteboard. The iOS pasteboard supports multiple representations simultaneously for an item on the pasteboard, so in this case, the pasteboard should hold both the image *and* its URL (e.g., for when pasting into text fields). This is consistent with Safari.

### CODE REVIEW

**Comment by thebnich** [comment]
Found one more issue: we don't validate the response, so `responseError` will be `nil` even for HTTP errors. One-line fix is to add a `validate()` line before handling the response.

**Comment by thebnich** [comment]
Nice catch. Ugh, as simple as it seems on the surface, `UIPasteboard` is one of the most convoluted, poorly-documented APIs I've run into on iOS. Added a few follow-up commits -- here's the process I used to get there:
 1. After some experimentation, I noticed that switching from `kUTTypeImage` to `kUTTypePNG` seems to magically work for all images I try (PNG, JPEG, and GIF). Not really clear what's going on here, but I'm guessing that `UIImagePNGRepresentation` is being called internally somewhere when giving `kUTTypePNG`. Even less clear is what we're *supposed* to do in this case, but this seems to work.
 1. Pasting the image into the message body in the Mail app, I noticed that it was pasting both the image *and* its URL. I tried using [`UIPasteboard setValue()`](https://developer.apple.com/library/prerelease/ios/documentation/UIKit/Reference/UIPasteboard_Class/index.html#//apple_ref/occ/instm/UIPasteboard/setValue:forPasteboardType:), instead, but that just did the same thing as setting `.image` (replaced the existing contents). So then I tried setting both the text and the image again using `.items`, and that finally seems to work as expected: it pastes only an image when supported, and falls back to text if needed.

 So in short, `UIPasteboard` has at least *four* ways to set the image: `.image`, `.addItems`, `.setValue`, and `.items`. What a cluster!@#$.

 1. I then noticed that when pasting into the **To:** field, I could paste the image URL copied from Firefox, but not from Safari. I found that setting the URL to `pasteboard.URL` instead of `pasteboard.string` fixes this. I'm guessing the **To:** field in the Mail app uses a certain kind of `UIKeyboardType`, and iOS is smart enough to know that it doesn't match a URL type.

After staring blankly at `UIPasteboard` for so long, I also noticed that it has a `changeCount` property that's updated each time the pasteboard is changed. I figured this was a slightly more proper condition to guard against rather than comparing the URL itself.

**Comment by thebnich** [comment]
Master: 60a3b28ee5e800ed884f19b6c90c9981fe2899be
v1.x: a0be117
